lib/script/script.js

```diff
@@ -1270,6 +1282,163 @@ class Script {
+            throw new ScriptError('INVALID_STACK_OPERATION', op, ip);
+
+          const v1 = stack.get(-2);
+          const v2 = stack.get(-1);
+          if (v1.length + v2.length > consensus.MAX_SCRIPT_PUSH) {
+            throw new ScriptError('PUSH_SIZE', op, ip);
+          }
+          stack.pop();
+          stack.pop();
+
+          stack.push(Buffer.concat([v1, v2]));
+
+          break;
+        }
+
+        case opcodes.OP_SPLIT: {
+          // (in position -- x1 x2)
+          if (stack.length < 2) {
+            throw new ScriptError('INVALID_STACK_OPERATION', op, ip);
+          }
+
+          const data = stack.get(-2);
+
+          // Make sure the split point is appropriate.
+          const pos = stack.getInt(-1, minimal, 4);
+          if (pos > data.length) {
+            throw new ScriptError('INVALID_SPLIT_RANGE', op, ip);
+          }
+
+          // Prepare the results in their own buffer as `data`
+          // will be invalidated.
+          const n1 = data.slice(0, pos);
+          const n2 = data.slice(pos);
+
+          // Replace existing stack values by the new values.
+          stack.set(-2, n1);
+          stack.set(-1, n2);
+
+          break;
+        }
+
+        //
+        // Bitwise logic
+        //
+        case opcodes.OP_AND:
+        case opcodes.OP_OR:
+        case opcodes.OP_XOR: {
+          // (x1 x2 - out)
+          if (stack.length < 2) {
+            throw new ScriptError('INVALID_STACK_OPERATION', op, ip);
+          }
+
+          const v1 = stack.get(-2);
+          const v2 = stack.get(-1);
+
+          // Inputs must be the same size
+          if (v1.length !== v2.length) {
+            throw new ScriptError('INVALID_OPERAND_SIZE', op, ip);
+          }
+
+          // To avoid allocating, we modify v1 in place.
+          switch (op.value) {
+            case opcodes.OP_AND:
+              for (let i = 1; i < v1.length; i++) {
+                v1[i] &= v2[i];
+              }
+              break;
+            case opcodes.OP_OR:
+              for (let i = 1; i < v1.length; i++) {
+                v1[i] |= v2[i];
+              }
+              break;
+            case opcodes.OP_XOR:
+              for (let i = 1; i < v1.length; i++) {
+                v1[i] ^= v2[i];
+              }
+              break;
+            default:
+              break;
+          }
+
+          // And pop v2.
+          stack.pop();
+
+          break;
+        }
+
+        //
+        // Conversion operations
+        //
+        case opcodes.OP_NUM2BIN: {
+          // (in size -- out)
+          if (stack.length < 2) {
+            throw new ScriptError('INVALID_STACK_OPERATION', op, ip);
+          }
+
+          const size = stack.getInt(-1, minimal, 4);
+          if (size > consensus.MAX_SCRIPT_PUSH) {
+            throw new ScriptError('PUSH_SIZE', op, ip);
+          }
+
+          stack.pop();
+
+          const n = stack.get(-1);
+          const v = ScriptNum.decode(n, true, size);
```

Comment:
Another thing to test will be the size: ScriptNum uses n64 internally but OP_NUM2BIN works on arbitrary sized NUM. So instead of depending on ScriptNum I think we need similar method that will operate directly on buffers without converting to numbers. Without assuming the length of the numbers.

Reply:
Well, after encoding it's limited to 520 Bytes but yeah, ScriptNum being n64 does overflow after 4 bytes. It's weird that OP_BIN2NUM only allows upto 4 bytes.